File: wasm/python-host/resolver_api.py
#!/usr/bin/env python3

import struct
from typing import List, Dict, Any

from google.protobuf import message
from datetime import datetime
from google.protobuf.timestamp_pb2 import Timestamp
from wasmtime import Engine, Store, Module, Instance, Func, Config, ValType, FuncType, Linker

# Import generated protobuf modules
from proto import messages_pb2
from proto.resolver import api_pb2

class ResolverApi:
    """Handles communication with the WASM module"""

    def __init__(self, wasm_bytes: bytes):
        # Create WASM engine and store
        # Create config and enable fuel consumption
        config = Config()
        # config.consume_fuel = True
        self.engine = Engine(config)
        self.store = Store(self.engine)

        # Compile the WASM module
        self.module = Module(self.engine, wasm_bytes)

        # Register host functions
        self._register_host_functions()

        # Get exported functions
        self.wasm_msg_alloc = self.instance.exports(self.store)["wasm_msg_alloc"]
        self.wasm_msg_free = self.instance.exports(self.store)["wasm_msg_free"]
        self.wasm_msg_guest_set_resolver_state = self.instance.exports(self.store)["wasm_msg_guest_set_resolver_state"]
        self.wasm_msg_guest_resolve = self.instance.exports(self.store)["wasm_msg_guest_resolve"]

    def _register_host_functions(self):
        """Register host functions that can be called from WASM"""

        def log_resolve(ptr: int) -> int:
            # Ignore payload; return Void
            response = messages_pb2.Response()
            response.data = messages_pb2.Void().SerializeToString()
            return self._transfer_response(response)

        def log_assign(ptr: int) -> int:
            # Ignore payload; return Void
            response = messages_pb2.Response()
            response.data = messages_pb2.Void().SerializeToString()
            return self._transfer_response(response)

        def current_time(ptr: int) -> int:
            """Host function to return current timestamp"""
            try:
                # Create timestamp
                timestamp = Timestamp()
                timestamp.FromDatetime(datetime.now())

                # Create response wrapper
                response = messages_pb2.Response()
                response.data = timestamp.SerializeToString()

                # Transfer response to WASM memory
                return self._transfer_response(response)
            except Exception as e:
                # Return error response
                error_response = messages_pb2.Response()
                error_response.error = str(e)
                return self._transfer_response(error_response)

        # # Register the host function
        # self.store.set_fuel(1000000)  # Add fuel for execution

        # Create function type: takes one i32 parameter, returns one i32
        func_type = FuncType([ValType.i32()], [ValType.i32()])
        host_func_time = Func(self.store, func_type, current_time)
        host_func_log_resolve = Func(self.store, func_type, log_resolve)
        host_func_log_assign = Func(self.store, func_type, log_assign)

        linker = Linker(self.store.engine)

        # Define the import with module and name
        linker.define(self.store, "wasm_msg", "wasm_msg_host_current_time", host_func_time)
        linker.define(self.store, "wasm_msg", "wasm_msg_host_log_resolve", host_func_log_resolve)
        linker.define(self.store, "wasm_msg", "wasm_msg_host_log_assign", host_func_log_assign)

        # Optional: current thread id function
        def current_thread_id() -> int:
            return 0
        linker.define(self.store, "wasm_msg", "wasm_msg_current_thread_id", Func(self.store, FuncType([], [ValType.i32()]), current_thread_id))

        # Instantiate the module with imports
        self.instance = linker.instantiate(self.store, self.module)

    def set_resolver_state(self, state: bytes) -> None:
        """Set the resolver state in the WASM module"""
        # Create request wrapper
        request = messages_pb2.Request()
        request.data = state
        # Transfer request to WASM memory - ensure it's a bytestring
        serialized_data: bytes = request.SerializeToString()
        req_ptr = self._transfer(serialized_data)
        # Call the WASM function
        results = self.wasm_msg_guest_set_resolver_state(self.store, req_ptr)
        resp_ptr = results
        # Consume the response
        self._consume_response(resp_ptr, lambda data: None)

    def resolve(self, request: api_pb2.ResolveFlagsRequest) -> api_pb2.ResolveFlagsResponse:
        """Resolve flags using the WASM module"""
        # Transfer request to WASM memory
        req_ptr = self._transfer_request(request)
        # Call the WASM function
        results = self.wasm_msg_guest_resolve(self.store, req_ptr)
        resp_ptr = results
        # Consume the response
        response = self._consume_response(resp_ptr, lambda data: self._parse_resolve_response(data))
        return response

    def _transfer_request(self, message: message.Message) -> int:
        """Transfer a protobuf message to WASM memory"""
        data = message.SerializeToString()
        request = messages_pb2.Request()
        request.data = data
        return self._transfer(request.SerializeToString())

    def _transfer_response(self, message: message.Message) -> int:
        """Transfer a response to WASM memory"""
        return self._transfer(message.SerializeToString())

    def _transfer(self, data: bytes) -> int:
        """Allocate memory in WASM and copy data"""
        # Allocate memory in WASM
        memory = self.instance.exports(self.store)["memory"]
        results = self.wasm_msg_alloc(self.store, len(data))
        # Write data to WASM memory
        memory.write(self.store, data, results)

        return results

    def _consume_response(self, addr: int, codec) -> Any:
        """Consume a response from WASM memory"""
        data = self._consume(addr)
        response = messages_pb2.Response()
        response.ParseFromString(data)

        if response.HasField('error'):
            raise Exception(f"WASM error: {response.error}")
        resp = codec(response.data)
        return resp

    def _consume(self, addr: int) -> bytes:
        """Read data from WASM memory and free it"""
        # Read length (assuming 4-byte length prefix)
        memory = self.instance.exports(self.store)["memory"]
        len_bytes = memory.read(self.store, addr - 4, addr)
        total_len = int.from_bytes(len_bytes, byteorder='little')
        length = total_len - 4

        # Read data
        data = memory.read(self.store, addr, addr + length)
        # Free memory
        self.wasm_msg_free(self.store, addr)

        return data

    def _parse_resolve_response(self, data: bytes) -> api_pb2.ResolveFlagsResponse:
        """Parse resolve response from protobuf data"""
        # Parse the protobuf bytes into ResolveFlagsResponse
        resolve_response = api_pb2.ResolveFlagsResponse()
        resolve_response.ParseFromString(data)
        return resolve_response

File: wasm/java-host/src/main/java/com/spotify/confidence/wasmresolvepoc/ResolverApi.java
package com.spotify.confidence.wasmresolvepoc;

import com.dylibso.chicory.annotations.WasmModuleInterface;
import com.dylibso.chicory.runtime.ExportFunction;
import com.dylibso.chicory.runtime.ImportFunction;
import com.dylibso.chicory.runtime.ImportValues;
import com.dylibso.chicory.runtime.Instance;
import com.dylibso.chicory.runtime.Memory;
import com.dylibso.chicory.wasm.WasmModule;
import com.dylibso.chicory.wasm.types.FunctionType;
import com.dylibso.chicory.wasm.types.ValType;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessage;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Timestamp;
import com.spotify.confidence.flags.resolver.v1.ResolveFlagsRequest;
import com.spotify.confidence.flags.resolver.v1.ResolveFlagsResponse;
import com.spotify.confidence.flags.resolver.v1.ResolvedFlag;

import rust_guest.Messages.SetResolverStateRequest;
import rust_guest.Messages;
import rust_guest.Messages.ResolveSimpleRequest;
import rust_guest.Types;

import java.util.List;
import java.util.function.Function;

@WasmModuleInterface(WasmResource.absoluteFile)
public class ResolverApi implements ResolverApi_ModuleImports, ResolverApi_WasmMsg {

  private static final FunctionType HOST_FN_TYPE = FunctionType.of(List.of(ValType.I32), List.of(ValType.I32));
  private final Instance instance;
  private final ResolverApi_ModuleExports exports;

  public ResolverApi() {
    instance = Instance.builder(ConfidenceResolver.load())
            .withImportValues(this.toImportValues())
            .withMachineFactory(ConfidenceResolver::create)
            .build();
    exports = new ResolverApi_ModuleExports(instance);
  }

  private Timestamp currentTime(Messages.Void unused) {
    return Timestamp.getDefaultInstance();
  }

  public void setResolverState(SetResolverStateRequest state) {
    final byte[] request = Messages.Request.newBuilder()
            .setData(state.toByteString())
            .build().toByteArray();
    int addr = transfer(request);
    int respPtr = exports.wasmMsgGuestSetResolverState(addr);
    consumeResponse(respPtr, Messages.Void::parseFrom);
  }

  public ResolveFlagsResponse resolve(ResolveFlagsRequest request) {
    int reqPtr = transferRequest(request);
    int respPtr = exports.wasmMsgGuestResolve(reqPtr);
    return consumeResponse(respPtr, ResolveFlagsResponse::parseFrom);
  }

  public ResolvedFlag resolve_simple(ResolveSimpleRequest request) {
    int reqPtr = transferRequest(request);
    int respPtr = exports.wasmMsgGuestResolveSimple(reqPtr);
    return consumeResponse(respPtr, ResolvedFlag::parseFrom);
  }

  private <T extends GeneratedMessage> T consumeResponse(int addr, ParserFn<T> codec) {
      try {
          Messages.Response response = Messages.Response.parseFrom(consume(addr));
          if(response.hasError()) {
            throw new RuntimeException(response.getError());
          } else {
            return codec.apply(response.getData().toByteArray());
          }
      } catch (InvalidProtocolBufferException e) {
          throw new RuntimeException(e);
      }
  }

  private <T extends GeneratedMessage> T consumeRequest(int addr, ParserFn<T> codec) {
    try {
      Messages.Request request = Messages.Request.parseFrom(consume(addr));
      return codec.apply(request.getData().toByteArray());
    } catch (InvalidProtocolBufferException e) {
      throw new RuntimeException(e);
    }
  }

  private int transferRequest(GeneratedMessage message) {
    final byte[] request = Messages.Request.newBuilder()
            .setData(message.toByteString())
            .build().toByteArray();
    return transfer(request);
  }

  private int transferResponseSuccess(GeneratedMessage response) {
    final byte[] wrapperBytes = Messages.Response.newBuilder()
            .setData(response.toByteString())
            .build().toByteArray();
    return transfer(wrapperBytes);
  }

  private int transferResponseError(String error) {
    final byte[] wrapperBytes = Messages.Response.newBuilder()
            .setError(error)
            .build().toByteArray();
    return transfer(wrapperBytes);
  }

  private byte[] consume(int addr) {
    final Memory mem = instance.memory();
    final int len = (int) (mem.readU32(addr - 4) - 4L);
    final byte[] data = mem.readBytes(addr, len);
    exports.wasmMsgFree(addr);
    return data;
  }

  private int transfer(byte[] data) {
    final Memory mem = instance.memory();
    int addr = exports.wasmMsgAlloc(data.length);
    mem.write(addr, data);
    return addr;
  }

  @Override
  public ResolverApi_WasmMsg wasmMsg() {
    return this;
  }

  @Override
  public int wasmMsgHostCurrentTime(int arg0) {
    try {
      return transferResponseSuccess(currentTime(Messages.Void.getDefaultInstance()));
    } catch (Exception e) {
      return transferResponseError(e.getMessage());
    }
  }

  private interface ParserFn<T> {

    T apply(byte[] data) throws InvalidProtocolBufferException;
  }

}
